src/utils.rs

```diff
@@ -154,34 +22,50 @@ where
 
 /// Returns error if index is not at a valid utf-8 char boundary
 #[inline]
-pub fn is_char_boundary<const N: usize>(s: &ArrayString<N>, idx: u8) -> Result<(), Utf8>
+pub fn is_char_boundary<const N: usize>(
+    s: &ArrayString<N>,
+    idx: impl Into<usize>,
+) -> Result<(), Utf8>
 where
     ArrayString<N>: ValidCapacity,
 {
+    let idx = idx.into();
     trace!("Is char boundary: {} at {}", s.as_str(), idx);
-    if s.as_str().is_char_boundary(idx.into()) {
+    if s.as_str().is_char_boundary(idx) {
         return Ok(());
     }
     Err(Utf8)
 }
 
+#[inline]
+pub unsafe fn is_char_boundary_at(arr: &[u8], index: usize) -> bool {
+    if index == 0 {
+        return true;
+    }
+    (*arr.get_unchecked(index) as i8) >= -0x40
+}
+
 /// Truncates string to specified size (ignoring last bytes if they form a partial `char`)
 #[inline]
-pub(crate) fn truncate_str(slice: &str, mut size: usize) -> &str {
-    trace!("Truncate str: {} at {}", slice, size);
+pub(crate) fn truncate_str(slice: &[u8], mut size: usize) -> &[u8] {
+    trace!(
+        "Truncate str: {} at {}",
+        unsafe { std::str::from_utf8_unchecked(slice) },
```

Comment:
I'm not sure I follow, why change slice to be &[u8] if you are going to treat as utf8?

I recommend using from_utf8 with error and printing debug in the trace at least

Reply:
printing bad utf8 will not crash, it will just print the unknown char char, so there is no reason to check it or else we would't see where the issue is.
The reason i changed to [u8] is because the uses of it already use [u8] at the point it is needed, and then need a [u8] anyway. Same reason as for the usize. Everything needs a usize in the standrad functions and it's already the most optimal value speed-wise.

Reply:
It seems that it used to be UB to simply cast invalid utf8 to &str, but this got removed. I'm not sure of the side-effects though. I would prefer still using at least _lossy to print. https://github.com/rust-lang/rust/issues/71033

But I can change that after my tests.

Your reasoning makes sense for changing the signature though.